### DISCUSSION TITLE
How to provide SourceAsset as input to an op

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#17718] [Q&A] [answered]
It’s unclear how to provide this runtime config:

```python
sourceUSEmployeeHandbook = SourceAsset(
    key=['llm', 'source', 'us_employee_handbook_pdf'], io_manager_key='s3_io_manager', description='US Employee Handbook', metadata={ 'bucket': 'our-data', 'key': 'my/path/test.pdf' })


@op(ins={'a': In(asset_key=AssetKey([ 'llm', 'source', 'us_employee_handbook_pdf' ]))})
def parse_document(context, a):
    context.log.info(a)
    
    return a
    
@graph
def my_graph():
    parse_document()
    
test_graph = my_graph.to_job()
```

I’m attempting to explicitly state that I want the value of `a` to be a source asset, which is directly defined above. When running the job, the UI instructs me to scaffold the missing configuration.  It’s unclear what this YAML is supposed to look like or how to provide the config via python.

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1699202378735469?thread_ts=1699202378.735469&cid=C01U954MEER).*

**Comment by jamiedemaria** [comment, marked as answer]
`@op`s don't automatically detect and provide upstream outputs like assets do. In order to use an asset or source asset as an input to an op, you'll need to pass it as a parameter. 

```python
sourceUSEmployeeHandbook = SourceAsset(
    key=['llm', 'source', 'us_employee_handbook_pdf'], io_manager_key='s3_io_manager', description='US Employee Handbook', metadata={ 'bucket': 'our-data', 'key': 'my/path/test.pdf' })


@op(ins={'a': In(asset_key=AssetKey([ 'llm', 'source', 'us_employee_handbook_pdf' ]))})
def parse_document(context, a):
    context.log.info(a)
    
    return a
    
@graph
def my_graph():
    parse_document(sourceUSEmployeeHandbook)  # this line is what changed
    
test_graph = my_graph.to_job()
```

Here is some additional documentation about this https://docs.dagster.io/concepts/ops-jobs-graphs/graphs#loading-an-asset-as-an-input

**Reply by seanlindo** [reply]
This is great, thank you! Last ask-- how can we represent this relationship in Dagit? As of now, there is no DAG showing the source asset flowing into the OP. 

Thank you so much!

**Reply by jamiedemaria** [reply]
I think this is not implemented in the UI right now. Can you open a GitHub issue so the UI team can address?
